shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt

```diff
@@ -0,0 +1,81 @@
+package org.hyperskill.app.core.data.repository_cache
+
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+class RepositoryCacheProxy<in Key : Any, Value : Any?>(
+    private val cache: RepositoryCache<Key, Value> =
+        InMemoryRepositoryCache(),
+    private val loadValuesFromRemote: suspend (keys: List<Key>) -> Result<List<Value>>,
+    private val getKey: (Value) -> Key?
+) {
+    private val mutex = Mutex()
+
+    suspend fun getValues(keys: List<Key>, force: Boolean): Result<List<Value>> {
+        if (keys.isEmpty()) return Result.success(emptyList())
+        return mutex.withLock {
+            if (keys.size == 1) {
+                // fast route for single key request
+                getSingleValue(keys.first(), force)
+            } else {
+                getManyValues(keys, force)
+            }
+        }
+    }
+
+    private suspend fun getSingleValue(key: Key, force: Boolean): Result<List<Value>> =
+        if (!force && cache.containsKey(key)) {
+            val cachedValue = cache[key]
+            if (cachedValue != null) {
+                Result.success(listOf(cachedValue))
+            } else {
+                loadManyValuesInternal(listOf(key))
+            }
+        } else {
+            loadManyValuesInternal(listOf(key))
+        }
+
+    private suspend fun getManyValues(keys: List<Key>, force: Boolean): Result<List<Value>> =
+        if (!force) {
+            val keysToLoadFromRemote: MutableList<Key> = mutableListOf()
+            val cachedValues: MutableList<Value> = mutableListOf()
+
+            keys.forEach { key ->
+                if (cache.containsKey(key)) {
+                    val cachedValue = cache[key]
+                    if (cachedValue != null) {
+                        cachedValues.add(cachedValue)
+                    } else {
+                        keysToLoadFromRemote.add(key)
+                    }
+                } else {
+                    keysToLoadFromRemote.add(key)
+                }
+            }
+
+            if (keysToLoadFromRemote.isEmpty()) {
+                Result.success(cachedValues)
+            } else {
+                loadManyValuesInternal(keysToLoadFromRemote)
+                    .map { remoteValues ->
+                        cachedValues + remoteValues
```

Comment:
@ivan-magda 
What is the reason to sort results by key?
We don't know which order is preferred for the user of the `RepositoryCacheProxy`. In my opinion, it is a user's responsibility to sort the results they want.

Reply:
Because the backend does this - when objects are requested by ids, the result objects will be sorted by the originally requested ids.

We may already rely on that backend behavior when implementing some features. Cache proxy will return result objects in unpredictable order, so that may lead to some bugs.

If we will not add sorting to the cache proxy, we must check current usages and remember in the future that result objects will not be sorted by ids.